### ISSUE TITLE
Incompatible LPWSTR typedef

### ISSUE THREAD

**Issue by shartte** [qmlnet/qmlnet#190]
This is more of a question/discussion than an actual bug.

Why is LPWSTR typedef'd to be char16_t rather than wchar_t?
I can guess why. wchar_t's size is implementation defined, but `UnmanagedType.LPWStr` _guarantees_ 2-byte per character.

The issue that arises from typedefing LPWSTR in QMLNet's headers is that if it is mixed with code that includes windows.h, it will not compile anymore, since wchar_t and char16_t are _NOT_ interchangeable:

```
qmlnet\src\native\QmlNet\QmlNet.h(12): error C2371: 'LPWSTR': redefinition; different basic types
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um\winnt.h(472): note: see declaration of 'LPWSTR'
```

If there's no other reason for it, I'd propose changing the native method declarations over to `const char16_t*` and removing the typedef. I am willing to make the PR for that ofcourse.

**Comment by pauldotknopf** [comment]
I don't have a problem with that. Either change everything to ```const char16_t*```, or create our own ```typedef``` that doesn't collide with ```windows.h```.

**Comment by shartte** [comment]
Just for completeness sake: I just checked QString::fromUtf16 again, and it says the following:
`This function is slow compared to the other Unicode conversions. Use QString(const QChar *, int) or QString(const QChar *) if possible.`

Does anyone know why that is?

**Comment by pauldotknopf** [comment]
Not me :)

**Comment by vadi2** [comment]
Thiago on #qt might know.

**Comment by shartte** [comment]
@pauldotknopf Thinking about this some more: Why don't we just use QChar* directly?
We don't actually do any conversion on the raw character data and it would save us the nasty reinterpret_cast that's otherwise needed.
Ofcourse the docs say "Most compilers treat it like an unsigned short." without actually specifying what "most" means. Although we might guard against this with a static_assert(sizeof(QChar) ==  2).
The `char16_t` constructor certainly makes it look like we can just use them interchangeably:

```
Q_DECL_CONSTEXPR QChar(char16_t ch) noexcept : ucs(ushort(ch)) {} // implicit
```

**Comment by pauldotknopf** [comment]
I'm not sure. I always thought ```QChar``` wasn't blittable with standard ```char16_t```, but it looks like I'm wrong.

**Comment by shartte** [comment]
Fixed by #203
